compiler/rustc_type_ir/src/outlives.rs

```diff
@@ -56,215 +55,170 @@ pub enum Component<I: Interner> {
             }
-        }
 
-        // OutlivesNominalType
-        ty::Slice(ty) |
-        ty::RawPtr(ty, _) => {
-            compute_components_for_ty(tcx, ty, out, visited);
-        }
-        ty::Tuple(tys) => {
-            for ty in tys.iter() {
-                compute_components_for_ty(tcx, ty, out, visited);
+            ty::Adt(_, _)
+            | ty::Foreign(_)
+            | ty::Array(_, _)
+            | ty::Pat(_, _)
+            | ty::Slice(_)
+            | ty::RawPtr(_, _)
+            | ty::Ref(_, _, _)
+            | ty::FnPtr(_)
+            | ty::Dynamic(_, _, _)
+            | ty::Tuple(_) => {
+                ty.super_visit_with(self);
             }
         }
+    }
 
-        // OutlivesReference
-        ty::Ref(lt, ty, _) => {
-            compute_components_for_lt(lt, out);
-            compute_components_for_ty(tcx, ty, out, visited);
+    fn visit_region(&mut self, lt: I::Region) -> Self::Result {
+        if !lt.is_bound() {
+            self.out.push(Component::Region(lt));
         }
+    }
 
-        ty::Dynamic(preds, lt, _) => {
-            compute_components_for_lt(lt, out);
-            for pred in preds.iter() {
-                match pred.skip_binder() {
-                    ty::ExistentialPredicate::Trait(tr) => {
-                        for arg in tr.args.iter() {
-                            compute_components_for_arg(tcx, arg, out, visited);
-                        }
-                    }
-                    ty::ExistentialPredicate::Projection(proj) => {
-                        for arg in proj.args.iter() {
-                            compute_components_for_arg(tcx, arg, out, visited);
-                        }
-                        match proj.term.kind() {
-                            ty::TermKind::Ty(ty) => {
-                                compute_components_for_ty(tcx, ty, out, visited)
-                            }
-                            ty::TermKind::Const(ct) => {
-                                compute_components_for_const(tcx, ct, out, visited)
-                            }
-                        }
-                    }
-                    ty::ExistentialPredicate::AutoTrait(..) => {}
-                }
-            }
+    fn visit_const(&mut self, ct: <I as Interner>::Const) -> Self::Result {
+        if !self.visited.insert(ct.into()) {
+            return;
         }
 
-        ty::FnPtr(sig) => {
-            for ty in sig.skip_binder().inputs_and_output.iter() {
-                compute_components_for_ty(tcx, ty, out, visited);
+        match ct.kind() {
+            ty::ConstKind::Param(_)
+            | ty::ConstKind::Bound(_, _)
+            | ty::ConstKind::Infer(_)
+            | ty::ConstKind::Placeholder(_)
+            | ty::ConstKind::Error(_) => {
+                // Trivial
+            }
+            ty::ConstKind::Expr(..) | ty::ConstKind::Value(..) | ty::ConstKind::Unevaluated(..) => {
+                ct.super_visit_with(self);
```

Comment:
2 ideas: change visited to only cache types. Change visit const to just be `super_visit_with` 🤔 this is the meaning of this match after all